crates/turborepo-lib/src/run/builder.rs

```diff
@@ -359,19 +364,34 @@ impl RunBuilder {
         let task_access = TaskAccess::new(self.repo_root.clone(), async_cache.clone(), &scm);
         task_access.restore_config().await;
 
-        let root_turbo_json = task_access
-            .load_turbo_json(&self.root_turbo_json_path)
-            .map_or_else(
-                || {
-                    TurboJson::load(
-                        &self.repo_root,
-                        &self.root_turbo_json_path,
-                        &root_package_json,
-                        is_single_package,
-                    )
-                },
-                Result::Ok,
-            )?;
+        let mut turbo_json_loader = if task_access.is_enabled() {
+            TurboJsonLoader::task_access(
+                self.repo_root.clone(),
+                self.root_turbo_json_path.clone(),
+                root_package_json.clone(),
+            )
+        } else if is_single_package {
+            TurboJsonLoader::single_package(
+                self.repo_root.clone(),
+                self.root_turbo_json_path.clone(),
+                root_package_json.clone(),
+            )
+        } else if self.allow_no_turbo_json && !self.root_turbo_json_path.exists() {
+            let package_scripts = workspace_package_scripts(pkg_dep_graph.packages());
+            // We explicitly set env mode to loose as otherwise no env vars will be passed
+            // to tasks.
+            self.opts.run_opts.env_mode = EnvMode::Loose;
```

Comment:
could we set this at the task level? I wanna keep `opts` as immutable as possible

Reply:
Doing this in another PR as adding this functionality to tasks isn't entirely trivial.